servlet.py

```diff
@@ -587,43 +594,173 @@ def urlRepl(self, base, attr, quote, aurl):
+        hsh = sha1(url.encode('utf-8')).hexdigest()
+        dirname = os.path.join(self.url_cache_path,
+                               # split it to avoid too many files in one dir:
+                               hsh[:1], hsh[1:2], hsh[2:])
+        return (dirname, "{}-{}".format(*pair))
+
+    def getCached(self, pair, url):
+        if not self.url_cache_path:
+            return None
+        if pair not in self.url_cache:
+            self.url_cache[pair] = {}
+        if url in self.url_cache[pair]:
+            logging.info("Got cache from memory")
+            return self.url_cache[pair][url]
+        dirname, basename = self.cachePath(pair, url)
+        path = os.path.join(dirname, basename)
+        if os.path.exists(path):
+            logging.info("Got cache on disk, we want to retranslate in background …")
+            with open(path, 'r') as f:
+                return (f.readline().strip(), f.read())
+
+    def retranslateCache(self, pair, url, cached):
+        """If we've got something from the cache, and it isn't in memory, then
+        it was from disk. We want to retranslate anything we found on
+        disk, since it's probably using older versions of the language
+        pair.
+
+        """
+        mem_cached = self.url_cache.get(pair, {}).get(url)
+        if mem_cached is None and cached is not None:
+            dirname, _ = self.cachePath(pair, url)
+            origpath = os.path.join(dirname, pair[0])
+            if os.path.exists(origpath):
+                return open(origpath, 'r').read()
+
+    def handleFetch(self, response):
+        if response.error is None:
+            return
+        elif response.code == 304:  # means we can use cache, so don't fail on this
+            return
+        else:
+            self.send_error(503, explanation="{} on fetching url: {}".format(response.code, response.error))
+
     @gen.coroutine
     def get(self):
         pair = self.getPairOrError(self.get_argument('langpair'),
                                    # Don't yet know the size of the text, and don't want to fetch it unnecessarily:
                                    -1)
-        if pair is not None:
-            pipeline = self.getPipeline(pair)
-            http_client = httpclient.AsyncHTTPClient()
-            url = self.get_argument('url')
-            request = httpclient.HTTPRequest(url=url,
-                                             # TODO: tweak
-                                             connect_timeout=20.0,
-                                             request_timeout=20.0)
-            response = yield http_client.fetch(request)
-            toTranslate = self.htmlToText(response.body, url)
+        if pair is None:
+            return
+        self.notePairUsage(pair)
+        mode_path = self.pairs['%s-%s' % pair]
+        markUnknown = self.get_argument('markUnknown', default='yes') in ['yes', 'true', '1']
+        url = self.get_argument('url')
+        if url.find('http://') == -1:
```

Comment:
Ohh yeah. Sorry for that. I could use `if not url.startswith('http'):`. If a string doesn't start with http, it obviously doesn't start with https. Does this make a valid check?